src/runtime_src/core/pcie/tools/cloud-daemon/aws/aws_dev.cpp

```diff
@@ -0,0 +1,655 @@
+    if (!d->isGood())
+        return 1;
+    return d->awsGetFirewall(resp, resp_len);
+}
+
+/*
+ * callback function that is used to handle MAILBOX_REQ_PEER_DATA msg
+ * kind DNA
+ *
+ * Input:
+ *        index: index of the FPGA device
+ * Output:
+ *        resp: response msg
+ *        resp_len: length of response msg 
+ * Return value:
+ *        0: success
+ *        1: failure
+ */
+int awsGetDna(size_t index, std::shared_ptr<struct xcl_dna> &resp,
+       size_t &resp_len)
+{
+    auto d = std::make_unique<AwsDev>(index, nullptr);
+    if (!d->isGood())
+        return 1;
+    return d->awsGetDna(resp, resp_len);
+}
+
+/*
+ * callback function that is used to handle MAILBOX_REQ_PEER_DATA msg
+ * kind SUBDEV
+ *
+ * Input:
+ *        index: index of the FPGA device
+ * Output:
+ *        resp: response msg
+ *        resp_len: length of response msg 
+ * Return value:
+ *        0: success
+ *        1: failure
+ */
+int awsGetSubdev(size_t index, std::shared_ptr<void> &resp,
+       size_t &resp_len)
+{
+    auto d = std::make_unique<AwsDev>(index, nullptr);
+    if (!d->isGood())
+        return 1;
+    return d->awsGetSubdev(resp, resp_len);
+}
+
+/*
+ * callback function that is used to handle MAILBOX_REQ_LOCK_BITSTREAM msg 
+ *
+ * Input:
+ *        index: index of the FPGA device
+ * Output:
+ *        none
+ * Return value:
+ *        0: success
+ *        1: failure
+ */
+int awsLockDevice(size_t index)
+{
+    auto d = std::make_unique<AwsDev>(index, nullptr);
+    if (!d->isGood())
+        return 1;
+    return d->awsLockDevice();
+}
+
+/*
+ * callback function that is used to handle MAILBOX_REQ_UNLOCK_BITSTREAM msg
+ * 
+ * Input:
+ *        index: index of the FPGA device
+ * Output:
+ *        none
+ * Return value:
+ *        0: success
+ *        1: failure
+ */
+int awsUnlockDevice(size_t index)
+{
+    auto d = std::make_unique<AwsDev>(index, nullptr);
+    if (!d->isGood())
+        return 1;
+    return d->awsUnlockDevice();
+}
+
+/*
+ * callback function that is used to handle MAILBOX_REQ_HOT_RESET msg
+ * 
+ * Input:
+ *        index: index of the FPGA device
+ * Output:
+ *        none
+ * Return value:
+ *        0: success
+ *        1: failure
+ */
+int awsResetDevice(size_t index)
+{
+    auto d = std::make_unique<AwsDev>(index, nullptr);
+    if (!d->isGood())
+        return 1;
+    return d->awsResetDevice();
+}
+
+/*
+ * callback function that is used to handle MAILBOX_REQ_RECLOCK msg
+ * 
+ * Input:
+ *        index: index of the FPGA device
+ *        obj: the mailbox req msg of type 'struct xclmgmt_ioc_freqscaling'
+ * Output:
+ *        none
+ * Return value:
+ *        0: success
+ *        1: failure
+ */
+int awsReClock2(size_t index, struct xclmgmt_ioc_freqscaling *&obj)
+{
+    auto d = std::make_unique<AwsDev>(index, nullptr);
+    if (!d->isGood())
+        return 1;
+    return d->awsReClock2(obj);
+}
+
+int AwsDev::awsLoadXclBin(const xclBin *&buffer)
```

Comment:
return code has to be used here, since it is content of the response of the mailbox.
return failure doesn't mean there is exception, instead it only mean the xbutil program not succeed.

Reply:
The comment of the calling function clearly states that the 1 is returned on error 0 on success, if so exceptions should be used as returns codes by default are ignored.  You can create your own exception class to capture the  return code if you need to know the value of the error code.

Reply:
Hi Soren,

Is the following what it should be?

try
{
    if (!d.isGood())
    {
         throw std::runtime_error("aws dev is not functional");
    }
}
catch(std::exception const& e)
{
    std::cout << "Exception: " << e.what() << "\n";
} 
return d.awsLoadXclBin()

Reply:
Nah, catch/throw within same function makes no sense, this doesn't change anything,   My point is that any return value that the caller *must* check to take corrective action is not C++'ish and should instead be an exception. 

If you prefer, please forget about the exception thing for now.  We can talk about this when I get back if you are interested.

Reply:
ok. I am going to keep it as-is and change to the exception way after talking to you later.